Add parameter to get_jacobian_matrix

felixvd opened the pull request:
### Description

I needed to expose the `reference_point_position` parameter of the `getJacobian` function in Python, so I added it as an optional parameter. Tested on a UR5, worked as expected.

Can be cherry-picked to kinetic and melodic just like #1501, where the `get_jacobian_matrix` was added that this PR modifies.

### Checklist
- [X] **Required by CI**: Code is auto formatted using [clang-format](http://moveit.ros.org/documentation/contributing/code)
- [X] Decide if this should be cherry-picked to other current ROS branches
- [X] While waiting for someone to review your request, please consider reviewing [another open pull request](https://github.com/ros-planning/moveit/pulls) to support the maintainers

felixvd commented:
Travis is returning this error, even after restarting the build:

```
<testsuite errors="1" failures="0" name="unittest.suite.TestSuite" tests="3" time="5.233"><testcase classname="__main__.PythonMoveGroupTest" name="test_get_jacobian_matrix" time="0.2506"><error type="ArgumentError">Python argument types in
    MoveGroupInterface.get_jacobian_matrix(MoveGroupInterface, list)
did not match C++ signature:
    get_jacobian_matrix(moveit::planning_interface::MoveGroupInterfaceWrapper {lvalue}, boost::python::list {lvalue}, boost::python::list {lvalue})
  File "/usr/lib/python2.7/unittest/case.py", line 329, in run
    testMethod()
  File "/root/ros_ws/src/moveit/moveit_ros/planning_interface/test/python_move_group.py", line 73, in test_get_jacobian_matrix
    result = self.group.get_jacobian_matrix(current)
```

But when I run `catkin run_tests --no-deps --this` in `moveit/moveit_ros/planning_interface`, I don't get any errors. The code is also running on our machine without problems. Anyone have an idea what's going on?

rhaschke commented:
It's written in the error: The call to get_jacobian_matrix in https://github.com/ros-planning/moveit/blob/master/moveit_ros/planning_interface/test/python_move_group.py#L73 is missing an argument.
This test was recently added in master. Probably you need to rebase your PR branch onto master to see this error locally as well.

rhaschke commented on moveit_commander/src/moveit_commander/move_group.py:
```diff
@@ -596,7 +596,7 @@ def retime_trajectory(self, ref_state_in, traj_in, velocity_scaling_factor=1.0,
         return traj_out
 
-    def get_jacobian_matrix(self, joint_values):
+    def get_jacobian_matrix(self, joint_values, reference_point_position=[0.0, 0.0, 0.0]):
```
It's bad practice to use mutable default arguments, see https://docs.python-guide.org/writing/gotchas.
Alternatively, use `None` as the default arg but forward the zero list if `None` was given.
```suggestion
    def get_jacobian_matrix(self, joint_values, reference_point_position=None):
```

felixvd replied:
Nice, thanks!

rhaschke commented on moveit_commander/src/moveit_commander/move_group.py:
```diff
@@ -596,7 +596,7 @@ def retime_trajectory(self, ref_state_in, traj_in, velocity_scaling_factor=1.0,
+    def get_jacobian_matrix(self, joint_values, reference_point_position=[0.0, 0.0, 0.0]):
         """ Get the jacobian matrix of the group as a list"""
-        return self._g.get_jacobian_matrix(joint_values)
+        return self._g.get_jacobian_matrix(joint_values, reference_point_position)
```
```suggestion
        return self._g.get_jacobian_matrix(joint_values, [0,0,0] if reference_point_position is None else reference_point_position)
```

rhaschke commented on moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:
```diff
@@ -536,14 +536,16 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
   }
 
-  Eigen::MatrixXd getJacobianMatrixPython(bp::list& joint_values)
+  Eigen::MatrixXd getJacobianMatrixPython(bp::list& joint_values, bp::list& reference_point_position)
```
To correctly handle the issue in https://github.com/ros-planning/moveit/pull/1595#issuecomment-519434052, reference_point_position needs to be optional here as well.
That's because this method can be called directly as well.

felixvd replied:
I don't know how to initialize a list to `[0.0, 0.0, 0.0]` for a default argument, but the new commit should work too. Thanks for taking another look.

rhaschke replied:
I implemented this.

rhaschke commented on moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:
```diff
@@ -536,14 +536,16 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
-    std::vector<double> v = py_bindings_tools::doubleFromList(joint_values);
+    std::vector<double> joint_vals = py_bindings_tools::doubleFromList(joint_values);
+    std::vector<double> ref_point_pos_vector = py_bindings_tools::doubleFromList(reference_point_position);
+    Eigen::Vector3d ref_point_pos(ref_point_pos_vector.data());
```
Use a `const Eigen::Map<Eigen::Vector3d>` wrapper here to avoid copying the data.

felixvd replied:
Can you explain more? `getJacobian` takes an `Eigen::Vector3d` as input, and I'm not sure how to avoid copying the data via `py_bindings_tools::doubleFromList`

rhaschke replied:
You copy twice: First `doubleFromList` is used to create a `std::vector`. Then, this std::vector is copied over to an `Eigen::Vector3d`. This second copy can be avoided using an Eigen::Map, which is essentially a wrapper around already allocated memory.

felixvd replied:
I missed this comment and am not at my PC anymore, but if simply replacing `Eigen::Vector3d` with `Eigen::Map<Eigen::Vector3d>` works, then I'll put it in on Monday.

rhaschke replied:
Done.

felixvd commented:
> It's written in the error: The call to get_jacobian_matrix in https://github.com/ros-planning/moveit/blob/master/moveit_ros/planning_interface/test/python_move_group.py#L73 is missing an argument.
> This test was recently added in master. Probably you need to rebase your PR branch onto master to see this error locally as well.

I saw the error message, but I don't understand it. The second argument is optional, so why would the call with one argument give an error?

I rebased on master and added your changes, but the tests succeeded locally for me at every point. We'll see if this commit changes anything.

rhaschke commented:
> The second argument is optional, so why would the call with one argument give an error?

We do have a two-layer wrapper here: There is the C++ level wrapper, and on top of this the thin python-level wrapper, which you adapted. Only the latter has the optional argument so far, not the former one. That's why my suggestion in https://github.com/ros-planning/moveit/pull/1595#discussion_r311939262.

rhaschke commented on moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:
```diff
@@ -536,11 +536,15 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
 
-  Eigen::MatrixXd getJacobianMatrixPython(bp::list& joint_values, bp::list& reference_point_position)
+  Eigen::MatrixXd getJacobianMatrixPython(const bp::list& joint_values,
+                                          const bp::list& reference_point_position = bp::list())
```
I suggest to use the default bp::object() here as well (if possible), which would be equivalent to None.
By default, boost::python only exposes the full signature. To also expose other variants (keeping the defaults), you need to perform [overload resolution](https://www.boost.org/doc/libs/1_66_0/libs/python/doc/html/tutorial/tutorial/functions.html#tutorial.functions.overloading).

felixvd commented:
I pushed a version with overload resolution because I didn't know how to check if `bp::object()` is `None`, and an additional test to confirm that the new parameter works.

> We do have a two-layer wrapper here: There is the C++ level wrapper, and on top of this the thin python-level wrapper, which you adapted. Only the latter has the optional argument so far, not the former one. That's why my suggestion in [#1595 (comment)](https://github.com/ros-planning/moveit/pull/1595#discussion_r311939262).

I understand that it's two layers, and when I first wrote this I put the default argument into both layers, but then I figured that the C++ level wrapper would not be called by anything but the Python wrapper. I also don't see where the test would be causing this error, and I don't understand why the test does not fail locally. Oh well. If this one passes I am happy.

felixvd commented on moveit_commander/src/moveit_commander/move_group.py:
```diff
@@ -596,7 +596,7 @@ def retime_trajectory(self, ref_state_in, traj_in, velocity_scaling_factor=1.0,
         return traj_out
 
-    def get_jacobian_matrix(self, joint_values, reference_point_position=None):
+    def get_jacobian_matrix(self, joint_values, reference_point=None):
```
For the record, I took the parameter name from the original function.

felixvd commented on moveit_ros/planning_interface/move_group_interface/src/wrap_python_move_group.cpp:
```diff
@@ -536,30 +536,28 @@ class MoveGroupInterfaceWrapper : protected py_bindings_tools::ROScppInitializer
-    state.setJointGroupPositions(group, joint_vals);
-    return state.getJacobian(group, ref_point_pos);
+    state.setJointGroupPositions(group, v);
+    return state.getJacobian(group, Eigen::Map<Eigen::Vector3d>(&ref[0]));
```
Thanks!
